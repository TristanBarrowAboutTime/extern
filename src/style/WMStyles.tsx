export default {
    color: {
        white: '#FFFFFF',
        black: '#000000',
        red: '#9B3E38',
        green: '#85B554',
        gray: {
            x_light: '#E9E9E9',
            light: '#E5E5E5',
            medium: '#A7AFB2',
            dark: '#777777',
            x_dark: '#4D4D4D',
            xx_dark: '#333333',
        },
        transparent: 'rgba(255,255,255,0)'
    },
    button: {
        paddingNormal: '8px 10px 10px 10px',
        paddingSmall: '6px 17px 4px 17px',
        shadow: '0 2px 4px #cccccc',
    },
    size: {
        xx_small: 2,
        x_small: 4,
        small: 8,
        medium: 16,
        large: 32,
        x_large: 48,
        xx_large: 64,
    },
    borderRadius: {
        standard: 4,
        circle: 100,
    },
    lineHeight: 20,
}