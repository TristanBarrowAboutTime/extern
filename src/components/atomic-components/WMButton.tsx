import React, { FunctionComponent } from 'react';
import styled from 'styled-components';
import WMStyles from '../../style/WMStyles';


/**
 * This will probably need to be turned into react-native at some point.
 */
export enum ButtonType {
    NORMAL = 'NORMAL',
    SMALL_GREEN = 'SMALL_GREEN',
    GREEN = 'GREEN',
    RED = 'RED',
    NAKED = 'NAKED',
    BARE = 'BARE'
}

type StyledButtonProps = {
    buttonType: ButtonType
    disabled: boolean
}

const getBgColor = (props: StyledButtonProps): string => {
    if (props.disabled) {
        switch(props.buttonType) {
            case ButtonType.NORMAL: 
            case ButtonType.SMALL_GREEN:
            case ButtonType.GREEN:  
            case ButtonType.RED:
                return WMStyles.color.gray.medium;
            case ButtonType.BARE:
            case ButtonType.NAKED:
                return 'rgba(0,0,0,0)';
        }
    } else {
        switch(props.buttonType) {
            case ButtonType.NORMAL: 
                return WMStyles.color.gray.xx_dark;
            case ButtonType.SMALL_GREEN:
            case ButtonType.GREEN:  
                return WMStyles.color.green;
            case ButtonType.RED:
                return WMStyles.color.red;
            case ButtonType.BARE:
            case ButtonType.NAKED:
                return 'rgba(0,0,0,0)';
            default: 
                return WMStyles.color.gray.xx_dark;
        }
    }
}

const getTextColor = (props: StyledButtonProps): string => {
    if (props.disabled) {
        switch(props.buttonType) {
            case ButtonType.NORMAL:
            case ButtonType.SMALL_GREEN:
            case ButtonType.GREEN:
            case ButtonType.RED:
                return WMStyles.color.white;
            case ButtonType.BARE:
            case ButtonType.NAKED:
                return WMStyles.color.gray.dark;
        }
    } else {
        switch(props.buttonType) {
            case ButtonType.NORMAL: 
            case ButtonType.SMALL_GREEN:
            case ButtonType.GREEN:  
            case ButtonType.RED:
                return WMStyles.color.white;
            case ButtonType.BARE:
            case ButtonType.NAKED:
                return WMStyles.color.gray.xx_dark;
            default: 
                return WMStyles.color.gray.xx_dark;
        }
    }
}

const getShadow = (props: StyledButtonProps): string => {
    if (props.disabled) return '0';
    if (props.buttonType === ButtonType.NAKED || 
        props.buttonType === ButtonType.BARE) return "0";
    else return `0 2px 4px #cccccc`;

}

const Button = styled.button`
    border: 0;
    border-radius: 4px;
    background-color: ${(props: StyledButtonProps) => getBgColor(props)};
    color: ${(props: StyledButtonProps) => getTextColor(props)};
    box-shadow: ${(props: StyledButtonProps) => getShadow(props)};
    padding: 10px;
    :hover {
        cursor: ${(props: StyledButtonProps) => props.disabled ? 'default' : 'pointer'}
    }
`;

export const WMButton: FunctionComponent<WMButtonProps> = ({
    buttonType = ButtonType.NORMAL, 
    text,
    onClick,
    styles = {},
    disabled = false
}: WMButtonProps) => {

    return (
        <Button 
            style={styles}
            onClick={onClick}
            disabled={disabled}
            buttonType={buttonType}
        >
            {text}
        </Button>
    );
}

type WMButtonProps = {
    buttonType?: ButtonType,
    text: string,
    styles?: object,
    onClick: () => void,
    disabled?: boolean,
}

export default WMButton;