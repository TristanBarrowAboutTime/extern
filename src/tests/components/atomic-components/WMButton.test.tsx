import React from 'react';
import renderer from 'react-test-renderer';
import WMButton, { ButtonType } from '../../../components/atomic-components/WMButton';
import 'jest-styled-components';
import WMStyles from '../../../style/WMStyles';

/*
 * for now snapshot testing should be sufficiant. 
 * it would be good to include click testing and className testing.
*/
describe('WMButton', () => {

    /**
     * Base
     */
    test('that button displays with minimal props', () => {
        const tree = renderer.create(
            <WMButton 
                text='Button'
                onClick={()=>{}}
            />
        );
        expect(tree).toMatchSnapshot();
        expect(tree).toHaveStyleRule('background-color', WMStyles.color.green);

    });

    /**
     * Normal
     */
    test('Normal buttons match snapshots', () => {
        const tree = renderer.create(
            <WMButton 
                text='Button'
                buttonType={ButtonType.NORMAL}
                onClick={()=>{}}
            />
        );
        expect(tree).toMatchSnapshot();
    });
    test('Normal buttons display disabled', () => {
        const tree = renderer.create(
            <WMButton 
                text='Button'
                buttonType={ButtonType.NORMAL}
                disabled={true}
                onClick={()=>{}}
            />
        );
        expect(tree).toMatchSnapshot();
    });


    /**
     * Green
     */
    test('Green buttons match snapshots', () => {
        const tree = renderer.create(
            <WMButton 
                text='Button'
                buttonType={ButtonType.GREEN}
                onClick={()=>{}}
            />
        );
        expect(tree).toMatchSnapshot();
    });
    test('Green buttons display disabled', () => {
        const tree = renderer.create(
            <WMButton 
                text='Button'
                buttonType={ButtonType.GREEN}
                disabled={true}
                onClick={()=>{}}
            />
        );
        expect(tree).toMatchSnapshot();
    });

    /**
     * Red
     */
    test('Red buttons match snapshots', () => {
        const tree = renderer.create(
            <WMButton 
                text='Button'
                buttonType={ButtonType.RED}
                onClick={()=>{}}
            />
        );
        expect(tree).toMatchSnapshot();
    });
    test('Red buttons display disabled', () => {
        const tree = renderer.create(
            <WMButton
                text='Button'
                buttonType={ButtonType.RED}
                disabled={true}
                onClick={()=>{}}
            />
        );
        expect(tree).toMatchSnapshot();
    });
    
    /**
     * Naked
     */
    test('Naked buttons match snapshots', () => {
        const tree = renderer.create(
            <WMButton 
                text='Button'
                buttonType={ButtonType.NAKED}
                onClick={()=>{}}
            />
        );
        expect(tree).toMatchSnapshot();
    });
    test('Naked buttons display disabled', () => {
        const tree = renderer.create(
            <WMButton 
                text='Button'
                buttonType={ButtonType.NAKED}
                disabled={true}
                onClick={()=>{}}
            />
        );
        expect(tree).toMatchSnapshot();
    });
});
