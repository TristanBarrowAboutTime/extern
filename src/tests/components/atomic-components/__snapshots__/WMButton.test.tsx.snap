// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WMButton Green buttons display disabled 1`] = `
.c0 {
  border: 0;
  border-radius: 4px;
  background-color: #A7AFB2;
  color: #FFFFFF;
  box-shadow: 0;
  padding: 10px;
}

.c0:hover {
  cursor: default;
}

<button
  className="c0"
  disabled={true}
  onClick={[Function]}
  style={Object {}}
>
  Button
</button>
`;

exports[`WMButton Green buttons match snapshots 1`] = `
.c0 {
  border: 0;
  border-radius: 4px;
  background-color: #85B554;
  color: #FFFFFF;
  box-shadow: 0 2px 4px #cccccc;
  padding: 10px;
}

.c0:hover {
  cursor: pointer;
}

<button
  className="c0"
  disabled={false}
  onClick={[Function]}
  style={Object {}}
>
  Button
</button>
`;

exports[`WMButton Naked buttons display disabled 1`] = `
.c0 {
  border: 0;
  border-radius: 4px;
  background-color: rgba(0,0,0,0);
  color: #777777;
  box-shadow: 0;
  padding: 10px;
}

.c0:hover {
  cursor: default;
}

<button
  className="c0"
  disabled={true}
  onClick={[Function]}
  style={Object {}}
>
  Button
</button>
`;

exports[`WMButton Naked buttons match snapshots 1`] = `
.c0 {
  border: 0;
  border-radius: 4px;
  background-color: rgba(0,0,0,0);
  color: #333333;
  box-shadow: 0;
  padding: 10px;
}

.c0:hover {
  cursor: pointer;
}

<button
  className="c0"
  disabled={false}
  onClick={[Function]}
  style={Object {}}
>
  Button
</button>
`;

exports[`WMButton Normal buttons display disabled 1`] = `
.c0 {
  border: 0;
  border-radius: 4px;
  background-color: #A7AFB2;
  color: #FFFFFF;
  box-shadow: 0;
  padding: 10px;
}

.c0:hover {
  cursor: default;
}

<button
  className="c0"
  disabled={true}
  onClick={[Function]}
  style={Object {}}
>
  Button
</button>
`;

exports[`WMButton Normal buttons match snapshots 1`] = `
.c0 {
  border: 0;
  border-radius: 4px;
  background-color: #333333;
  color: #FFFFFF;
  box-shadow: 0 2px 4px #cccccc;
  padding: 10px;
}

.c0:hover {
  cursor: pointer;
}

<button
  className="c0"
  disabled={false}
  onClick={[Function]}
  style={Object {}}
>
  Button
</button>
`;

exports[`WMButton Red buttons display disabled 1`] = `
.c0 {
  border: 0;
  border-radius: 4px;
  background-color: #A7AFB2;
  color: #FFFFFF;
  box-shadow: 0;
  padding: 10px;
}

.c0:hover {
  cursor: default;
}

<button
  className="c0"
  disabled={true}
  onClick={[Function]}
  style={Object {}}
>
  Button
</button>
`;

exports[`WMButton Red buttons match snapshots 1`] = `
.c0 {
  border: 0;
  border-radius: 4px;
  background-color: #9B3E38;
  color: #FFFFFF;
  box-shadow: 0 2px 4px #cccccc;
  padding: 10px;
}

.c0:hover {
  cursor: pointer;
}

<button
  className="c0"
  disabled={false}
  onClick={[Function]}
  style={Object {}}
>
  Button
</button>
`;

exports[`WMButton that button displays with minimal props 1`] = `
.c0 {
  border: 0;
  border-radius: 4px;
  background-color: #333333;
  color: #FFFFFF;
  box-shadow: 0 2px 4px #cccccc;
  padding: 10px;
}

.c0:hover {
  cursor: pointer;
}

<button
  className="c0"
  disabled={false}
  onClick={[Function]}
  style={Object {}}
>
  Button
</button>
`;
